### PR TITLE
Employee certification service refactor

### PR DESCRIPTION
**Pull request by KAYEMBAJONATHAN**
- Add refactor employee certification service 
- Add Update unit test
- Add mock new unit test to cover all the line
Screenshot 2024-07-25 073740

### CODE REVIEW

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 1]
```diff
@@ -9,26 +10,45 @@ namespace HRIS.Services.Services;
...
+        var exists = await CheckIfCertificationExists(employeeCertificationDto.EmployeeId);
+        if (!exists)
+            throw new CustomException("Certificate not found");
 
-        if (!await CheckEmployeeExists(employeeCertificationDto.EmployeeId))
-            throw new CustomException("Employee not found");
+        if (_identity.Role is not ("SuperAdmin" or "Admin" or "Talent" or "Journey") && employeeCertificationDto.Id != _identity.EmployeeId)
+        {
+            throw new CustomException("Unauthorized access.");
+        }
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        if (!_identity.IsSupport && employeeCertificationDto.Id != _identity.EmployeeId)
            throw new CustomException("Unauthorized access.");
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 2]
```diff
@@ -59,23 +79,29 @@ public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(i
...
-        if (!await CheckEmployeeExists(employeeCertificationDto.EmployeeId))
-            throw new CustomException("Employee not found");
+        var exists = await CheckIfCertificationExists(employeeCertificationDto.EmployeeId);
+        if (!exists)
+            throw new CustomException("Certificate not found");
+
+        if (_identity.Role is not ("SuperAdmin" or "Admin" or "Talent" or "Journey") && employeeCertificationDto.Id != _identity.EmployeeId)
+        {
+            throw new CustomException("Unauthorized access.");
+        }
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        if (!_identity.IsSupport && employeeCertificationDto.Id != _identity.EmployeeId)
            throw new CustomException("Unauthorized access.");
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 3]
```diff
@@ -59,23 +79,29 @@ public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(i
...
 
         return (await _db.EmployeeCertification.Update(new EmployeeCertification(employeeCertificationDto))).ToDto();
     }
 
-    public async Task<EmployeeCertificationDto> DeleteEmployeeCertification(int id)
+    public async Task<EmployeeCertificationDto> DeleteEmployeeCertification(int Id)
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
    public async Task<EmployeeCertificationDto> DeleteEmployeeCertification(int id)
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 4]
```diff
@@ -59,23 +79,29 @@ public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(i
...
     {
-        return (await _db.EmployeeCertification.Delete(id)).ToDto();
-    }
+        var exists = await CheckIfCertificationExists(Id);
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        var exists = await CheckIfCertificationExists(id);
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 5]
```diff
@@ -59,23 +79,29 @@ public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(i
...
 
-    public async Task<bool> CheckEmployeeExists(int employeeId)
-    {
-        var employee = await _db.Employee
-                                .Get(employee => employee.Id == employeeId)
-                                .FirstOrDefaultAsync();
+        if (_identity.Role is not ("SuperAdmin" or "Admin" or "Talent" or "Journey") && Id != _identity.EmployeeId)
+        {
+            throw new CustomException("Unauthorized access.");
+        }
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        if (!_identity.IsSupport && id != _identity.EmployeeId)
            throw new CustomException("Unauthorized access.");
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 6]
```diff
@@ -59,23 +79,29 @@ public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(i
...
 
-        return employee != null;
+        return (await _db.EmployeeCertification.Delete(Id)).ToDto();
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        return (await _db.EmployeeCertification.Delete(id)).ToDto();
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 7]
```diff
@@ -41,13 +61,13 @@ public async Task<EmployeeCertificationDto> GetEmployeeCertification(int employe
...
-        
+
         return employeeCertification;
     }
 
     public async Task<List<EmployeeCertificationDto>> GetAllEmployeeCertifications(int employeeId)
     {
-        if (!await CheckEmployeeExists(employeeId))
+        if (!await CheckIfEmployeeExists(employeeId))
             throw new CustomException("Employee not found");
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        if (!_identity.IsSupport && employeeId != _identity.EmployeeId)
            throw new CustomException("Unauthorized access.");
```

File: HRIS.Services/Services/EmployeeCertificationService.cs [hunk 8]
```diff
@@ -9,26 +10,45 @@ namespace HRIS.Services.Services;
...
 
     public async Task<EmployeeCertificationDto> GetEmployeeCertification(int employeeId, int certificationId)
     {
-        if (!await CheckEmployeeExists(employeeId))
+        if (!await CheckIfEmployeeExists(employeeId))
             throw new CustomException("Employee not found");
 
+        if(!await CheckIfCertificationExists(certificationId))
+            throw new CustomException("Certificate not found");
+
```
Comment (bseptember) [inline HRIS.Services/Services/EmployeeCertificationService.cs]:
```suggestion
        if (!_identity.IsSupport && employeeId != _identity.EmployeeId)
            throw new CustomException("Unauthorized access.");
```

**Review by bseptember — CHANGES_REQUESTED** [review]
@KAYEMBAJONATHAN you need to add two additional unit tests to check identity role. 
for SaveEmployeeCertification & GetEmployeeCertification

File: HRIS.Services.Tests/Services/EmployeeCertificationServiceUnitTests.cs [hunk 9]
```diff
@@ -1,6 +1,10 @@
 ﻿using System.Linq.Expressions;
+using HRIS.Models;
 using HRIS.Services.Interfaces;
```
Comment (EduanO) [inline HRIS.Services.Tests/Services/EmployeeCertificationServiceUnitTests.cs:3]:
```suggestion
```

File: HRIS.Services/Services/EmployeeCertificationService.cs
Comment (EduanO) [inline HRIS.Services/Services/EmployeeCertificationService.cs:1]: Please fix naming conventions on methods

`SaveEmployeeCertification` -> `CreateEmployeeCertification`
`GetEmployeeCertification` -> `GetEmployeeCertificationByEmployeeIdAndCertificationId`
`GetAllEmployeeCertifications` -> `GetEmployeeCertificationsByEmployeeId`

**Review by EduanO — CHANGES_REQUESTED** [review]

**Comment by KAYEMBAJONATHAN** [comment]
Screenshot 2024-07-26 105338

**Review by bseptember — APPROVED** [review]

**Review by nzikalala — APPROVED** [review]

**Review by EduanO — APPROVED** [review]

**Pull request merged**